File: 0061-rotate-list/0061-rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function (head, k) {
    function findlength(head) {
        let length = 0;
        let current = head;
        let prev;
        while (current != null) {
            length++;
            prev = current;
            current = current.next
        }
        return [length, prev];

    }
    let tail = findlength(head)[1]
    const n = findlength(head)[0];
    k = k % n;
    if (k == 0) {
        return head;
    }
    if (head==null || head.next==null){
        return head
    }
    let oldhead = head;
    let newhead = head;
    let newtail = null;
    for (let i = 0; i < n - k; i++) {
        newtail= newhead;
        newhead = newhead.next;
    }
    newtail.next = null;
    tail.next = oldhead;
    return newhead



};

File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int Length(ListNode* head) {
        int l = 0;
        while (head) {
            l++;
            head = head->next;
        }
        return l;
    }
    ListNode* rotateRight(ListNode* head, int k) {

        int n = Length(head);
        if (n == 0 || k==0)
            return head;
        k = k % n;
        if (k == 0)
            return head;
        ListNode* current = head;
        ListNode* prev = NULL;
        ListNode* Next = head;
        for (int i = 0; i < k+1; i++) {
            current = current->next;
            prev = Next;
            Next = Next->next;
        }
        for (int i = 0; i < n-2-k; i++) {
            current = current->next;
        }
        current->next = head;
        prev->next = NULL;
        return Next;

    }
};